src/dinit-iostream.cc

```diff
@@ -0,0 +1,744 @@
+    }
+    return r;
+}
+
+bool ostream::write_nx(const char *msg) noexcept
+{
+    size_t len = strlen(msg);
+    return (put(msg, len) == static_cast<ssize_t>(len));
+}
+
+bool ostream::write(const char *msg)
+{
+    bool r = write_nx(msg);
+    if (!r) {
+        // This will definitely throw an exception
+        throw_exception_on(io_states::buffer_failbit | io_states::io_failbit);
+    }
+    return r;
+}
+
+bool ostream::write_nx(char msg) noexcept
+{
+    return (put(&msg, 1) == 1);
+}
+
+bool ostream::write(char msg)
+{
+    bool r = write_nx(msg);
+    if (!r) {
+        // This will definitely throw an exception
+        throw_exception_on(io_states::buffer_failbit | io_states::io_failbit);
+    }
+    return r;
+}
+
+bool ostream::write_nx(const std::string &msg) noexcept
+{
+    return (put(msg.c_str(), msg.size()) == static_cast<ssize_t>(msg.size()));
+}
+
+bool ostream::write(const std::string &msg)
+{
+    bool r = write_nx(msg);
+    if (!r) {
+        // This will definitely throw an exception
+        throw_exception_on(io_states::buffer_failbit | io_states::io_failbit);
+    }
+    return r;
+}
+
+bool ostream::write_nx(const int num) noexcept
+{
+    char tmp[INT_MAX_CHARS + 1];
+    int len = snprintf(tmp, INT_MAX_CHARS + 1, "%d", num);
+    return (put(tmp, len) == len);
+}
+
+bool ostream::write(const int num)
+{
+    bool r = write_nx(num);
+    if (!r) {
+        // This will definitely throw an exception
+        throw_exception_on(io_states::buffer_failbit | io_states::io_failbit);
+    }
+    return r;
+}
+
+bool ostream::write_nx(const unsigned num) noexcept
+{
+    char tmp[UINT_MAX_CHARS + 1];
+    int len = snprintf(tmp, UINT_MAX_CHARS + 1, "%d", num);
+    return (put(tmp, len) == len);
+}
+
+bool ostream::write(const unsigned num)
+{
+    bool r = write_nx(num);
+    if (!r) {
+        // This will definitely throw an exception
+        throw_exception_on(io_states::buffer_failbit | io_states::io_failbit);
+    }
+    return r;
+}
+
+bool ostream::write_nx(const long num) noexcept
+{
+    char tmp[LONG_MAX_CHARS + 1];
+    int len = snprintf(tmp, LONG_MAX_CHARS + 1, "%ld", num);
+    return (put(tmp, len) == len);
+}
+
+bool ostream::write(const long num)
+{
+    bool r = write_nx(num);
+    if (!r) {
+        // This will definitely throw an exception
+        throw_exception_on(io_states::buffer_failbit | io_states::io_failbit);
+    }
+    return r;
+}
+
+bool ostream::write_nx(const unsigned long num) noexcept
+{
+    char tmp[ULONG_MAX_CHARS + 1];
+    int len = snprintf(tmp, ULONG_MAX_CHARS + 1, "%ld", num);
+    return (put(tmp, len) == len);
+}
+
+bool ostream::write(const unsigned long num)
+{
+    bool r = write_nx(num);
+    if (!r) {
+        // This will definitely throw an exception
+        throw_exception_on(io_states::buffer_failbit | io_states::io_failbit);
+    }
+    return r;
+}
+
+bool ostream::write_nx(const endline &) noexcept
+{
+    int r = put("\n", 1);
+    r += flush_nx();
```

Comment:
Please change this to be more idiomatic (don't rely on conversion of bool to int, and don't heap conditions together). Eg

    int r = put("\n", 1);
    if (r != 1) return false;
    if (flush_nx()) return false;
    return true;

Reply:
(as currently written it is actually wrong, it will return true even if `flush_nx` fails. My example code above is correct I think, after the edit I just made).

Reply:
I wrote it like this
```c++
    int r = put("\n", 1);
    if (r != 1) return false;
    return (flush_nx());
```